File: app/src/main/java/social/entourage/android/authentication/ComplexPreferences.java
package social.entourage.android.authentication;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.text.TextUtils;

import com.google.gson.Gson;

import java.lang.reflect.Type;

/**
 * Source : https://github.com/fsilvestremorais/android-complex-preferences/blob/master/ComplexPreferences/src/br/com/kots/mob/complex/preferences/ComplexPreferences.java
 */
public class ComplexPreferences {

    private static ComplexPreferences complexPreferences;
    private SharedPreferences preferences;
    private SharedPreferences.Editor editor;
    private static Gson GSON = new Gson();

    @SuppressLint("CommitPrefEdits")
    private ComplexPreferences(Context context, String namePreferences, int mode) {
        if (TextUtils.isEmpty(namePreferences)) {
            namePreferences = "complex_preferences";
        }
        preferences = context.getSharedPreferences(namePreferences, mode);
        editor = preferences.edit();
    }

    public static ComplexPreferences getComplexPreferences(Context context, String namePreferences, int mode) {
        if (complexPreferences == null) {
            complexPreferences = new ComplexPreferences(context, namePreferences, mode);
        }
        return complexPreferences;
    }

    public void putObject(String key, Object object) {
        if(object == null) {
            editor.putString(key, "");
        }
        if(TextUtils.isEmpty(key)) {
            throw new IllegalArgumentException("key is empty or null");
        }
        editor.putString(key, GSON.toJson(object));
    }

    public void commit() {
        editor.commit();
    }

    public <T> T getObject(String key, Class<T> a) {
        String gson = preferences.getString(key, null);
        if (gson == null) {
            return null;
        } else {
            try{
                return GSON.fromJson(gson, a);
            } catch (Exception e) {
                throw new IllegalArgumentException("Object storage with key " + key + " is instanceof other class");
            }
        }
    }

    public <T> T getObject(String key, Type t) {
        String gson = preferences.getString(key, null);
        if (gson == null) {
            return null;
        } else {
            try{
                return GSON.fromJson(gson, t);
            } catch (Exception e) {
                throw new IllegalArgumentException("Object storage with key " + key + " is instanceof other class: "+e.getLocalizedMessage());
            }
        }
    }
}

File: app/src/main/java/social/entourage/android/authentication/ComplexPreferences.kt
package social.entourage.android.authentication

import android.content.Context
import android.content.SharedPreferences
import com.google.gson.Gson
import java.lang.reflect.Type

/**
 * Source : https://github.com/fsilvestremorais/android-complex-preferences/blob/master/ComplexPreferences/src/br/com/kots/mob/complex/preferences/ComplexPreferences.java
 */
class ComplexPreferences(context: Context, namePreferences: String, mode: Int) {
    private val preferences: SharedPreferences
    private val editor: SharedPreferences.Editor

    fun putObject(key: String, `object`: Any?) {
        require(key.isNotEmpty()) { "key is empty or null" }
        editor.putString(key, if (`object` == null) "" else GSON.toJson(`object`))
    }

    fun commit() {
        editor.commit()
    }

    fun apply() {
        editor.apply()
    }

    fun <T> getObject(key: String, a: Class<T>?): T? {
        val gson = preferences.getString(key, null) ?: return null
        return try {
            GSON.fromJson(gson, a)
        } catch (e: Exception) {
            null
        }
    }

    fun <T> getObjectFromType(key: String, t: Type): T? {
        val gson = preferences.getString(key, null) ?: return null
        return try {
            GSON.fromJson(gson, t)
        } catch (e: Exception) {
            null
        }
    }

    companion object {
        private val GSON = Gson()
    }

    init {
        val safeNamePreferences = namePreferences.ifEmpty { "complex_preferences" }
        preferences = context.getSharedPreferences(safeNamePreferences, mode)
        editor = preferences.edit()
    }
}